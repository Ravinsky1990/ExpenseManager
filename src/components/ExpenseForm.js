import React from "react";
import  moment from "moment";
import {SingleDatePicker} from "react-dates";


export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props)

        this.state = {
            description: props.expense ? props.expense.description : "",
            note: props.expense ? props.expense.note : "",
            amount: props.expense ? (props.expense.amount/100).toString() : "",
            createAt: props.expense ? moment(props.expense.createAt): moment(),
            calendarfocused: false,
            error: ""
        }
    }
    
    onDescriptionChange = (ev) =>{
        const description = ev.target.value;
        this.setState(()=>({description}))
    };
    onNoteChange = (ev)=>{
        const note = ev.target.value
        this.setState(()=>({note}))
    };
    onAmountChage = (ev)=>{
        const amount = ev.target.value;
        if(amount.match(/^\d*(\.\d{0,2})?$/)){
            this.setState(()=>({amount}))
        }
    }
    onDateChange = (createAt) =>{
        if(createAt){
            this.setState(()=>({createAt}))
        }
    }
    onFocusChange= ({focused})=>{
        this.setState(()=>({calendarfocused:focused}))
    }
    onSubmit = (ev)=>{
        ev.preventDefault();
        if(!this.state.description || !this.state.amount){
            //set error masege
            this.setState(()=>({error:"Please provide description and amount"}))
        }
        else{
            //cleare error
            this.setState(()=>({error:""}))
            this.props.onSubmit({
                description:this.state.description,
                amount: parseFloat(this.state.amount,10)*100,
                createAt: this.state.createAt.valueOf(),
                note: this.state.note

            })
        }
    }

    render(){
        return (
            <div>
                <p>{this.state.error}</p>
                <form onSubmit={this.onSubmit}>
                    <input type="text"
                    placeholder="Description"
                    autoFocus value={this.state.description}
                    onChange={this.onDescriptionChange} />
                    
                    <input type="text"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChage}
                     />

                     <SingleDatePicker
                     date = {this.state.createAt }
                     onDateChange ={this.onDateChange}
                     focused ={this.state.calendarfocused}
                     onFocusChange ={this.onFocusChange}
                     numberOfMonths = {1}
                     isOutsideRange = {()=>false}
                     />
                    <textarea placeholder="Add a note to your expense (optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                    ></textarea>

                    <button>Edd Expense</button>
                </form>
            </div>
        )
    }
}