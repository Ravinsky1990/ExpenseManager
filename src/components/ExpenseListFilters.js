import React from "react";
import {connect} from "react-redux";
import {DateRangePicker} from "react-dates";
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from "../actions/filtres";

export class ExpenseListFilters extends React.Component{
    state ={
        calendarFocused:null
    };
    onDatesChange = ({startDate, endDate}) =>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    };
    onFocusChange = (calendarFocused) =>{
        this.setState(()=>({
            calendarFocused
        }))
    }
    onTextCange = (ev)=>{
        this.props.setTextFilter(ev.target.value)
    }
    onSortChange = (ev)=>{
        if (ev.target.value == "date"){
            this.props.sortByDate()
        }
        else if (ev.target.value == "amount"){
            this.props.sortByAmount()
        }
    }


    render(){
        return(
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextCange} />
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                startDate ={this.props.filters.startDate}
                endDate ={this.props.filters.endDate}
                onDatesChange ={this.onDatesChange}
                focusedInput ={this.state.calendarFocused}
                onFocusChange = {this.onFocusChange}
                showClearDates = {true}
                numberOfMonths = {1}
                isOutsideRange = {()=> false}
                
                />
            </div>
        )
        
    }
}

//connect to store

const mapStateToProps =(state)=>{
    return{
        filters:state.filters
    }
}

const mapDispatchToProps = (dispatch)=>({
    setTextFilter: (text)=>dispatch(setTextFilter(text)),
    sortByDate: ()=>dispatch(sortByDate()),
    sortByAmount: ()=>dispatch(sortByAmount()),
    setStartDate: (startDate)=> dispatch(setStartDate(startDate)),
    setEndDate: (endDate)=>dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)