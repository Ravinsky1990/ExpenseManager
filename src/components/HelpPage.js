import React from "react";

const HelpPage = () => (
    <div>
        HelpPage component
    </div>
)

export default HelpPage;