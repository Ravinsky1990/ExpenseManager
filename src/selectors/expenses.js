import moment from "moment";

//Visible expenses

export default (expenses, {text, sortBy, startDate, endDate})=>{
    return expenses.filter((expense)=>{
        const createAtmoment = moment(expense.createAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createAtmoment, "day") : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createAtmoment, "day") : true; 
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b)=>{
        if (sortBy === "date"){
            return a.createAt < b.createAt ? 1 : -1;
        } else if(sortBy ==="amount"){
            return a.amount < b.amount ? 1 : -1;
        }
    });
}
