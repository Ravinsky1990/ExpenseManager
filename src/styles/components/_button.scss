.button{
    background: $blue;
    border: none;
    display: inline-block;
    padding: $s-size;
    font-size: 1.8rem;
    line-height: 1;
    font-weight: 300;
    color: white;
    text-decoration: none;
}

.button--link{
    background: none;
    
}