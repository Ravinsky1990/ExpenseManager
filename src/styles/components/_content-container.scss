.content-container{
    margin: auto;
    padding: 0 $m-size;
    max-width: 80rem;
}