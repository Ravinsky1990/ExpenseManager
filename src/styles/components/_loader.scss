.loader{
    display: flex;
    align-items: center;
    justify-content: center;
    height: 100vh;
    width: 100vw;
}

.loader__image{
    height: 6rem;
    width: 6rem;
}