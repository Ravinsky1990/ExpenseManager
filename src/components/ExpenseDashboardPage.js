import React from "react";
import ExpenseListFilters from "./ExpenseListFilters";

import ExpenseList from "./ExpenseList";

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
)

export default ExpenseDashboardPage;