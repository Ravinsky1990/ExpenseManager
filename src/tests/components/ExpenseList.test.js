import React from "react";
import {shallow} from "enzyme";
import {ExpenseList} from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";

test("should render Expenselist with expenses", ()=>{
    const wrapper = shallow(<ExpenseList expenses={expenses}/>)
    expect(wrapper).toMatchSnapshot()
})

test("should render Expenselist with empty message", ()=>{
    const wrapper = shallow(<ExpenseList expenses={[]}/>)
    expect(wrapper).toMatchSnapshot()
})