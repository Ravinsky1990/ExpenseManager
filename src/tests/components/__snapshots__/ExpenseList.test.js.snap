// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expenselist with empty message 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`should render Expenselist with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={200}
    createAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={30000}
    createAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={40000}
    createAt={345600000}
    description="Credit card"
    id="3"
    key="3"
    note=""
  />
</div>
`;
